etl/scripts/faostat/create_new_snapshots.py

```diff
@@ -0,0 +1,377 @@
+            "publication_year": self.publication_year,
+            "publication_date": str(self.publication_date),
+            "date_accessed": VERSION,
+            "version": VERSION,
+            "url": f"{FAO_DATA_URL}/{self._dataset_metadata['DatasetCode']}",
+            "source_data_url": self.source_data_url,
+            "file_extension": "zip",
+            "license_url": LICENSE_URL,
+            "license_name": LICENSE_NAME,
+            "is_public": True,
+        }
+
+    def to_snapshot(self):
+        """
+        Create a snapshot.
+
+        Download the dataset from the source, create the corresponding metadata file and create the snapshot.
+        """
+        # Create metadata file for current domain dataset.
+        create_snapshot_metadata_file(self.metadata)
+
+        # URI of snapshot to be created.
+        snapshot_uri = f"{self.metadata['namespace']}/{self.metadata['version']}/{self.metadata['short_name']}.{self.metadata['file_extension']}"
+
+        # Create a new snapshot.
+        snap = Snapshot(snapshot_uri)
+
+        # Download data from source.
+        snap.download_from_source()
+
+        # Add file to DVC and upload to S3.
+        snap.dvc_add(upload=True)
+
+
+def load_faostat_catalog() -> List[Dict[str, Any]]:
+    datasets = requests.get(FAO_CATALOG_URL).json()["Datasets"]["Dataset"]
+    return datasets
+
+
+def is_dataset_already_up_to_date(
+    existing_snapshots: List[Snapshot], source_data_url: str, source_modification_date: dt.date
+) -> bool:
+    """Check if our latest snapshot for a particular domain dataset is already up-to-date.
+
+    Iterate over all snapshots in the catalog and check if:
+    * The URL of the source data coincides with the one of the current domain dataset.
+    * The last time the source data was accessed was more recently than the source's last modification date.
+
+    If those conditions are fulfilled, we consider that the current dataset does not need to be updated.
+
+    Args:
+        source_data_url (str): URL of the source data.
+        source_modification_date (date): Last modification date of the source dataset.
+    """
+    dataset_up_to_date = False
+    for snapshot in existing_snapshots:
+        snapshot_source_data_url = snapshot.metadata.source_data_url
+        snapshot_date_accessed = parser.parse(str(snapshot.metadata.date_accessed)).date()
+        if (snapshot_source_data_url == source_data_url) and (snapshot_date_accessed > source_modification_date):
+            dataset_up_to_date = True
+
+    return dataset_up_to_date
+
+
+class FAOAdditionalMetadata:
+    def __init__(self):
+        # Assign current date to additional metadata.
+        self.publication_date = dt.datetime.today().date()
+        self.publication_year = self.publication_date.year
+
+        # Initialise the combined metadata to be downloaded from FAOSTAT.
+        self.faostat_metadata = None
+
+    @property
+    def metadata(self):
```

Comment:
(Methods are missing type hints, but we don't enforce them anymore so up to you.)

Reply:
I basically adapted the code we used to have in walden ingest, so I didn't put much additional effort. But I'll try to add type hints now.